tests/casper/langs/test_state_lang.py

```diff
@@ -0,0 +1,292 @@
+# protocols with a single genesis block
+GENESIS_PROTOCOLS = [BlockchainProtocol]
+# each val starts with different block
+INITAL_MESSAGE_PROTOCOLS = [BinaryProtocol, OrderProtocol, IntegerProtocol]
+
+
+@pytest.mark.parametrize(
+    'protocol',
+    (
+        OrderProtocol,
+        BlockchainProtocol,
+        BinaryProtocol,
+        IntegerProtocol
+    )
+)
+def test_init(protocol, test_weight):
+    state_lang = StateLanguage(test_weight, protocol, False)
+
+    assert isinstance(state_lang.plot_tool, protocol.PlotTool)
+    assert len(state_lang.validator_set) == len(test_weight)
+    assert state_lang.validator_set.validator_weights() == set(test_weight.values())
+    assert state_lang.validator_set.weight() == sum(test_weight.values())
+
+    assert callable(state_lang.make_message)
+    assert callable(state_lang.make_invalid)
+    assert callable(state_lang.send_message)
+    assert callable(state_lang.plot)
+
+
+@pytest.mark.parametrize(
+    'protocol',
+    (
+        OrderProtocol,
+        BlockchainProtocol,
+        BinaryProtocol,
+        IntegerProtocol
+    )
+)
+def test_init_creates_network(protocol, test_weight):
+    state_lang = StateLanguage(test_weight, protocol, False)
+
+    assert isinstance(state_lang.network, Network)
+
+
+
+@pytest.mark.parametrize(
+    'protocol',
+    (
+        OrderProtocol,
+        BlockchainProtocol,
+        BinaryProtocol,
+        IntegerProtocol
+    )
+)
+def test_init_validators_have_only_inital_messages(protocol, test_weight):
+    state_lang = StateLanguage(test_weight, protocol, False)
+
+    for validator in state_lang.network.validator_set:
+        assert len(validator.view.justified_messages) == 1
+
+
+@pytest.mark.parametrize(
+    'test_string, error',
+    [
+        ('M0-A', None),
+        ('M0-A S1-A', None),
+        ('M0-A M0-B', None),
+        ('M0-A M0-B S1-A S1-B', None),
+        ('I1-B', NotImplementedError),
+        ('A-B', ValueError),
+        ('BA0-A, S1-A', KeyError),
+        ('BA0-A S1-A', KeyError),
+        ('RR0-A-A', ValueError),
+        ('B0-A S1-A T1-A', KeyError),
+        ('RR0-AB1-A', ValueError),
+        ('RRR', ValueError),
+        ('A0-A S1-A', KeyError),
+    ]
+)
+def test_parse_only_valid_tokens(test_string, test_weight, error):
+    state_lang = StateLanguage(test_weight, BinaryProtocol, False)
+
+    if isinstance(error, type) and issubclass(error, Exception):
+        with pytest.raises(error):
+            state_lang.parse(test_string)
+        return
+
+    state_lang.parse(test_string)
+
+
+@pytest.mark.parametrize(
+    'test_strings, error',
+    [
+        (['M0-A', 'S1-A'], None),
+        (['M0-A', 'M0-B S1-A S1-B'], None),
+        (['M0-A', 'I1-B'], NotImplementedError),
+        (['M0-A', 'S1-A T1-A'], KeyError),
+        (['M1-A', 'RR0-AB1-A'], ValueError),
+    ]
+)
+def test_parse_only_valid_tokens_split_strings(test_strings, test_weight, error):
+    for protocol in PROTOCOLS:
```

Comment:
Can possibly build the parametrize data with a loop and then pass in "protocol, test_strings, error". Will help out if there happens to be a test broken against one protocol but not the others.

Reply:
Same for tests below. If it's too messy, don't worry about it.